ci: split CI into separate files

Molkree opened the pull request:
- closes #1309

What happens here:
1. One big CI file split into 3 smaller ones. Testing is still pretty big with some duplication between jobs.
2. Some steps slightly reordered (getting the date followed immediately by getting cached DB for the date, etc).
3. Same step names everywhere.
4. Install `wheel` where needed to avoid `legacy 'setup.py install'` during tool installation.
5. Doc build/CVE scan will run on the latest Python version (I have removed explicit version specification).
6. Removed `matrix.os` from testing jobs because we have separate jobs for Ubuntu and Windows so each one had just one entry in matrix.
7. Pulled `ACTIONS` env value to the workflow level as it is being set in all jobs. Same for `LONG_TEST` env value but it is being overwritten in the `long_tests` job.
8. Long tests will now run when we change `cli.py` file. Maybe we should do this for all `.py` changes or on schedule? I'm open to suggestions.
9. Windows tests ran inside conda due to the `pdftotext` dependency. This was particularly bad because when we thought we were running on Python 3.10 we did in fact just run on Python 3.9 because conda still didn't have 3.10 support (until recently). I've moved conda activation (and PDF tests) to the end of Win job so it won't interfere until the last moment.
    - We also specify Python version in conda now, meaning that Python 3.9 job will actually use 3.9 in conda and 3.10 job will use 3.10 in conda.
10. You can now just check/uncheck to run long tests when starting tests workflow manually: 
![image](https://user-images.githubusercontent.com/12860284/151211686-aa7b25fe-a004-4eb4-a41f-527c4c8d6e41.png)
11. Bumped [technote-space/get-diff-action](https://github.com/technote-space/get-diff-action) (used to determine if we need to run long tests) and codecov/codecov-action ([v1 will be sunset on February 1 2022](https://github.com/codecov/codecov-action#%EF%B8%8F--deprecration-of-v1)).

Molkree commented:
Long tests are failing in my fork with some errors in idna and some missing files, no idea yet if it's from I did or unrelated...
[Check run](https://github.com/Molkree/cve-bin-tool/runs/4954767809?check_suite_focus=true)
```log
FAILED test/test_checkers.py::TestCheckerVersionParser::test_filename_is[curl-libcurl.so.4-expected_results3]
FAILED test/test_checkers.py::TestCheckerVersionParser::test_filename_is[bzip2-bzip2.so-expected_results1]
```
```log
    def test_filename_is(self, checker_name, file_name, expected_results):
        """Test a checker's filename detection"""
        checkers = pkg_resources.iter_entry_points("cve_bin_tool.checker")
        for checker in checkers:
            if checker.name == checker_name:
>               Checker = checker.load()

test/test_checkers.py:119: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/pkg_resources/__init__.py:2464: in load
    self.require(*args, **kwargs)
/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/pkg_resources/__init__.py:2487: in require
    items = working_set.resolve(reqs, env, installer, extras=self.extras)
/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/pkg_resources/__init__.py:785: in resolve
    new_requirements = dist.requires(req.extras)[::-1]
/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/pkg_resources/__init__.py:2749: in requires
    dm = self._dep_map
/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/pkg_resources/__init__.py:3033: in _dep_map
    self.__dep_map = self._compute_dependencies()
/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/pkg_resources/__init__.py:3042: in _compute_dependencies
    for req in self._parsed_pkg_info.get_all('Requires-Dist') or []:
/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/pkg_resources/__init__.py:3024: in _parsed_pkg_info
    metadata = self.get_metadata(self.PKG_INFO)
/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/pkg_resources/__init__.py:1412: in get_metadata
    value = self._get(path)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pkg_resources.PathMetadata object at 0x7fdd6d26ed30>
path = '/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/idna-3.3.dist-info/METADATA'

    def _get(self, path):
>       with open(path, 'rb') as stream:
E       FileNotFoundError: [Errno 2] No such file or directory: '/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/idna-3.3.dist-info/METADATA'
```

terriko commented:
Sorry, merging the bandit stuff caused a conflict, but I think it should be not too hard conceptually to resolve by pulling the bandit changes into the linter file instead of pythonapp.yml

Those errors aren't occurring on main right now (having just merged the bandit stuff I can see that I just got a clean build). If I had to guess, the test file it's using isn't getting created/downloaded properly because things happen in a slightly different order.  Maybe rather than removing pythonapp.yml you could try keeping the stuff that went into testing.yml in the original filename so we can see the diff better here, and that might help us track down the problem.  Can always rename the file in a later PR, unless it turns out that pythonapp.yml has some magic in GitHub Actions that makes this work (unlikely, but I wouldn't rule it out entirely)

Molkree commented:
Closing this until I fix long tests, just don't want it to run needlessly on all my experimental commits.

Molkree commented:
Alright, it didn't take long, looks like my first guess was the correct one.

Two tests were failing in long tests job:
```
FAILED test/test_checkers.py::TestCheckerVersionParser::test_filename_is[curl-libcurl.so.4-expected_results3]
FAILED test/test_checkers.py::TestCheckerVersionParser::test_filename_is[bzip2-bzip2.so-expected_results1]
```
How it was before:
```yml
      # steps installing dependencies thru apt-get and pip
      - name: try single cli run of tool  # O1
        run: |
          python -m pip install -e .
          NO_EXIT_CVE_NUM=1 python -m cve_bin_tool.cli test/assets -n json -u latest
      - name: Run async tests  # O2
        run: >
          pytest --cov --cov-append -n 4 -v
          # skipping some tests here
```
How I made it:
```yml
      # steps installing dependencies thru apt-get
      - name: Install pdftotext and cve-bin-tool  # N1
        run: |
          python -m pip install --upgrade pip
          python -m pip install --upgrade setuptools
          python -m pip install --upgrade wheel
          python -m pip install --upgrade pdftotext
          python -m pip install --upgrade .
      - name: Try single CLI run of tool  # N2
        run: |
          NO_EXIT_CVE_NUM=1 python -m cve_bin_tool.cli test/assets -n json -u latest
      - name: Run async tests  # N3
        run: >
          pytest --cov --cov-append -n 4 -v
          # skipping some tests here
```

Albeit some rearranging of steps, the difference is in how I install cve-bin-tool itself. Previously it was an editable install (`pip install -e .` in step O1) and I changed it to a simple local install (`pip install --upgrade .` in step N1).

I don't understand why it breaks these tests. I don't understand why it only fails on 2 files out of 52 in this test. And finally, I don't understand why it doesn't fail in normal (not long) and Windows tests (where I've also replaced `--editable` with `--upgrade`). Any ideas are highly appreciated!

terriko reviewed (approved):
Wow, I also have many questions about this.

If I had to guess : `--editable` just grabs the whole darned directory and plunks a symlink or something down, but `--upgrade`  is only copying the files that setup tells it to copy into the environment.   So we've either misconfigured setup.py and are missing something, or those tests passing at all is a fluke because of some leftover file?  I don't like either of those.  And that's all speculation off the top of my head.  But we do know that `--editable` causes weird behaviours (see #1323) that we haven't resolved so there's definitely something to look into here.

That said, I think this is a huge improvement to our CI setup files and I'd like to get it merged sooner rather than later so we can take advantage of it.  So... I guess I'm going to approve this and open a separate bug so we can figure out what's up with the install problem?
